---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# fastreeR

<!-- badges: start -->
<!-- badges: end -->

The goal of fastreeR is to provide functions for calculating distance matrix,
building phylogenetic tree or performing hierarchical clustering 
between samples, directly from a VCF or FASTA file.

## Installation

To install `fastreeR` package:
```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("fastreeR")
```

You can install the development version of fastreeR like so:
```{r, eval=FALSE}
devtools::install_github("gkanogiannis/fastreeR")
```

## Memory requirements for VCF input

At minimum, make sure to allocate for JVM at least 10 bytes per variant 
per sample. If there are \code{n} samples and \code{m} variants 
allocate 10xnxm bytes of RAM. 
For example, for processing a VCF file containing data 
for 1 million variants and 1 thousand samples, allocate at least :
10^6 x 10^3 x 10 = 10^10 bytes = 10GB of RAM.
For optimal execution, allocate more RAM than minimum. This will trigger 
less times garbage collections and hence less pauses.

In order to allocate 3GB of RAM for the JVM, through R code, use:
```{r, eval=FALSE}
options(java.parameters="-Xmx3G")
```

A rough estimation for the required RAM, if sample and variant numbers 
are not known, is half the size of the uncompressed VCF file. For example 
for processing a VCF file, which uncompressed occupies 2GB of disk space, 
allocate 1GB of RAM.


## Distances from VCF

Calculates a cosine type dissimilarity measurement between the
\code{n} samples of a VCF file.

Biallelic or multiallelic (maximum 7 alternate alleles) SNP and/or INDEL 
variants are considered, phased or not. Some VCF encoding examples are:

 \itemize{
     \item heterozygous variants : \code{1/0} or \code{0/1} or \code{0/2}
                                    or \code{1|0} or \code{0|1} or \code{0|2}
     \item homozygous to the reference allele variants : \code{0/0} 
                                                        or \code{0|0}
     \item homozygous to the first alternate allele variants : \code{1/1} 
                                                                or \code{1|1}
 }

If there are \code{n} samples and \code{m} variants, an \code{nxn}
zero-diagonal symmetric distance matrix is calculated.
The calculated cosine type distance (1-cosine_similarity)/2 is in the range
[0,1] where value 0 means completely identical samples (cosine is 1), 
value 0.5 means perpendicular samples (cosine is 0) 
and value 1 means completely opposite samples (cosine is -1).

The calculation is performed by a Java backend implementation, 
that supports multi-core CPU utilization
and can be demanding in terms of memory resources.
By default a JVM is launched with a maximum memory allocation of 512 MB.
When this amount is not sufficient,
the user needs to reserve additional memory resources,
before loading the package,
by updating the value of the \code{java.parameters} option.
For example in order to allocate 4GB of RAM,
the user needs to issue \code{options(java.parameters="-Xmx4g")}
before \code{library(fastreeR)}.

Output file will contain \code{n+1} lines.
The first line contains the number \code{n} of samples
and number \code{m} of variants, separated by space.
Each of the subsequent \code{n} lines contains \code{n+1} values,
separated by space.
The first value of each line is a sample name
and the rest \code{n} values
are the calculated distances of this sample to all the samples.
Example output file of the distances of 3 samples calculated from 1000 variants:
\tabular{llll}{
    3 1000 \tab \cr
    Sample1 \tab 0.0 \tab 0.5 \tab 0.2\cr
    Sample2 \tab 0.5 \tab 0.0 \tab 0.9\cr
    Sample3 \tab 0.2 \tab 0.9 \tab 0.0\cr
}

## Distances from FASTA



## Tree from distances



## Clusters from tree


