---
title: "fastreeR Vignette"
author: 
-   name: "Anestis Gkanogiannis"
    email: anestis@gkanogiannis.com
package: fastreeR
output: BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{fastreeR}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

# About fastreeR

The goal of fastreeR is to provide functions for calculating distance matrix,
building phylogenetic tree or performing hierarchical clustering 
between samples, directly from a VCF or FASTA file.

# Installation

To install `fastreeR` package:
```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("fastreeR")
```

# Preparation

## Allocate RAM and load required libraries
You should allocate minimum 10kb per sample per variant of RAM for the JVM.
The more RAM you allocate, the faster the execution will be (less pauses 
for garbage collection).
```{r, eval=TRUE}
options(java.parameters="-Xmx512M")
library(fastreeR)
library(utils)
library(ape)
library(stats)
library(grid)
```

## Download sample vcf file
We download, in a temporary location, a small vcf file 
from 1K project, with around 150 samples and 100k variants (SNPs and INDELs).
```{r, eval=TRUE}
temp_vcf_url <-
    paste0("http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/",
        "1000_genomes_project/release/20190312_biallelic_SNV_and_INDEL/",
        "supporting/related_samples/",
        "ALL.chrX.shapeit2_integrated_snvindels_v2a_related_samples_27022019.",
        "GRCh38.phased.vcf.gz")
temp_vcf <- tempfile(fileext = ".vcf.gz")
utils::download.file(url = temp_vcf_url, destfile = temp_vcf, method = "curl")
```

## Download sample fasta files
We download, in temporary location, 5 small bacterial genomes
and we merge them into a single fasta file.
```{r, eval=TRUE}
temp_fasta_urls <- c(
    paste0("https://ftp.ncbi.nih.gov/genomes/archive/old_refseq/Bacteria/",
            "Mycobacterium_liflandii_128FXT_uid59005/NC_011355.fna"),
    paste0("https://ftp.ncbi.nih.gov/genomes/archive/old_refseq/Bacteria/",
            "Pelobacter_propionicus_DSM_2379_uid58255/NC_008607.fna"),
    paste0("https://ftp.ncbi.nih.gov/genomes/archive/old_refseq/Bacteria/",
            "Rickettsia_prowazekii_Chernikova_uid158053/NC_017049.fna"),
    paste0("https://ftp.ncbi.nih.gov/genomes/archive/old_refseq/Bacteria/",
            "Salmonella_enterica_serovar_Newport_SL254_uid58831/NC_009140.fna"),
    paste0("https://ftp.ncbi.nih.gov/genomes/archive/old_refseq/Bacteria/",
            "Staphylococcus_aureus_JH1_uid58457/NC_009619.fna")
)
temp_fastas <- lapply(seq_len(5), function(x) tempfile(fileext = ".fna"))
for (i in seq(1,5)) {
    utils::download.file(url = temp_fasta_urls[i], 
                            destfile = temp_fastas[[i]], method="curl")    
}

temp_fasta <- tempfile(fileext = ".fasta")
for (i in temp_fastas){
    x <- readLines(i)
    write(x, file = temp_fasta, append = TRUE)
}
```

# Functions on vcf files

## Sample Statistics
```{r, fig.cap = "Sample statistics from vcf file", fig.wide = TRUE, echo=TRUE}
my.vcf.istats <- fastreeR::vcf2istats(inputfile = temp_vcf)
plot(my.vcf.istats[,7:9])
```

## Calculate distances from vcf
The most time consuming process is calculating distances between samples.
Assign more processors in order to speed up this operation.
```{r, eval=TRUE}
my.vcf.dist <- fastreeR::vcf2dist(inputfile = temp_vcf, threads = 1)
```

## Histogram of distances
```{r, fig.cap="Histogram of distances from vcf file", fig.wide=TRUE, echo=TRUE}
my.vcf.hist.png <- tempfile(fileext = ".png")
my.vcf.hist <- fastreeR::dist2hist(input = my.vcf.dist, 
                                    outputfile = my.vcf.hist.png,
                                    settings = "100,640,480")
grid::grid.raster(my.vcf.hist)
```
We note 150 zero distance values, as it should be expected (vcf file had 50 
samples) and we also note two distinct groups of distances. One around of 
distance value 0.055 and the second around distance value 0.065.

## Plot tree from fastreeR::dist2tree
Notice that the generated tree is ultrametric.
```{r, fig.cap = "Tree from vcf with fastreeR", fig.wide = TRUE, echo = TRUE}
my.vcf.tree <- fastreeR::dist2tree(input.dist = my.vcf.dist)
plot(ape::read.tree(text = my.vcf.tree), direction = "down", cex = 0.3)
ape::add.scale.bar()
ape::axisPhylo(side = 2)
```

Of course the same can be achieved directly from the vcf file, 
without calculating distances.
```{r, fig.cap="Tree from vcf with fastreeR",fig.wide=TRUE,echo=TRUE,eval=FALSE}
my.vcf.tree <- fastreeR::vcf2tree(inputfile = temp_vcf)
plot(ape::read.tree(text = my.vcf.tree), direction = "down", cex = 0.3)
ape::add.scale.bar()
ape::axisPhylo(side = 2)
```
As expected from the histogram of distances, two groups of samples also 
emerge in the tree. The two branches, one at height around 0.055 and the second 
around height 0.065, are clearly visible.

## Plot tree from stats::hclust
For comparison, we generate a tree by using \code{stats} package and distances
calculated by \code{fastreeR}.
```{r, fig.cap = "Tree from vcf with stats::hclust", fig.wide = TRUE, echo=TRUE}
my.vcf.tree.stats <- stats::hclust(my.vcf.dist)
plot(my.vcf.tree.stats, ann = FALSE, cex = 0.3)
```
Although it does not initially look very similar, because it is not ultrametric,
it is indeed quite the same tree. We note again the two groups (two branches) 
of samples and the 4 samples, possibly clones, that they show very close 
distances between them.
 
## Hierarchical Clustering
We can identify the two groups of samples, apparent from the hierarchical tree,
by using \code{\link[fastreeR]{dist2clusters}} 
or \code{\link[fastreeR]{vcf2clusters}} 
or \code{\link[fastreeR]{tree2clusters}}.
By playing a little with the \code{cutHeight} parameter, we find that a
value of \code{cutHeight=0.067} cuts the tree into two branches.
The first group contains 106 samples and the second 44.
```{r, eval=TRUE}
my.vcf.clust <- fastreeR::dist2clusters(input.dist = my.vcf.dist, 
                                    cutHeight = 0.067)
tree <- my.vcf.clust[[1]]
clusters1 <- my.vcf.clust[[2]][[2]][1]
clusters2 <- my.vcf.clust[[2]][[2]][2]
clusters1
clusters2
```

# Functions on fasta files

Similar analysis we can perform when we have samples represented as 
sequences in a fasta file.

## Calculate distances from fasta
```{r, eval=TRUE}
my.fasta.dist <- fastreeR::fasta2dist(inputfile = temp_fasta, kmer = 6)
```

## Histogram of distances
```{r, fig.cap="Histogram of distances from fasta file",fig.wide=TRUE,echo=TRUE}
my.fasta.hist.png <- tempfile(fileext = ".png")
my.fasta.hist <- fastreeR::dist2hist(input = my.fasta.dist, 
                                    outputfile = my.fasta.hist.png,
                                    settings = "100,640,480")
grid::grid.raster(my.fasta.hist)
```

## Plot tree from fastreeR::dist2tree
```{r, fig.cap = "Tree from fasta with fastreeR", fig.wide = TRUE, echo = TRUE}
my.fasta.tree <- fastreeR::dist2tree(input.dist = my.fasta.dist)
plot(ape::read.tree(text = my.fasta.tree), direction = "down", cex = 0.3)
ape::add.scale.bar()
ape::axisPhylo(side = 2)
```

## Plot tree from stats::hclust
```{r, fig.cap = "Tree from fasta with stats::hclust", fig.wide=TRUE, echo=TRUE}
my.fasta.tree.stats <- stats::hclust(my.fasta.dist)
plot(my.fasta.tree.stats, ann = FALSE, cex = 0.3)
```

# SessionInfo
```{r setup}
utils::sessionInfo()
```
